back/app/services/phase_insights_service.rb

```diff
@@ -0,0 +1,298 @@
+class PhaseInsightsService
+  SUPPORTED_CATEGORICAL_FIELD_TYPES = %w[select checkbox multiselect].freeze
+
+  def insights_data(phase)
+    cached_insights_data(phase)
+  end
+
+  private
+
+  def initialize
+    @permissions_custom_fields_service = Permissions::PermissionsCustomFieldsService.new
+  end
+
+  # TODO: Implement caching? (may not be needed if performance good enough)
+  # Removed funky caching used when originally used Singleton pattern.
+  def cached_insights_data(phase)
+    visitors_data = VisitsService.new.phase_visitors_data(phase)
+    participations = phase.pmethod.participations
+    flattened_participations = participations.values.flatten
+    participant_ids = flattened_participations.pluck(:participant_id).uniq
+
+    metrics_data(phase, participations, participant_ids, visitors_data).merge(
+      demographics: { fields: demographics_data(phase, flattened_participations, participant_ids) }
+    )
+  end
+
+  def metrics_data(phase, participations, participant_ids, visitors_data)
+    base_metrics = base_metrics(participations, participant_ids, visitors_data)
+
+    participation_method_metrics = {
+      phase.participation_method => participation_method_metrics(phase, participations)
+    }
+
+    { metrics: base_metrics.merge(participation_method_metrics) }
+  end
+
+  # TODO: Add user_hash to all participations where available (e.g. ideas, comments),
+  # and use in the various calculations of unique users,
+  # whenever user_id is nil (for non-logged in users, participating anonymously), etc.
+  def base_metrics(participations, participant_ids, visitors_data)
+    total_participant_count = participant_ids.count
+    flattened_participations = participations.values.flatten
+    participants_last_7_days_count = flattened_participations.select { |p| p[:acted_at] >= 7.days.ago }.pluck(:participant_id).uniq.count
+
+    {
+      visitors: visitors_data[:total],
+      visitors_last_7_days: visitors_data[:last_7_days],
+      participants: total_participant_count,
+      participants_last_7_days: participants_last_7_days_count,
+      engagement_rate: visitors_data[:total] > 0 ? (total_participant_count.to_f / visitors_data[:total]).round(3) : 0
+    }
+  end
+
+  # TBD to cover the different needs for different participation methods
+  def participation_method_metrics(phase, participations)
+    case phase.participation_method
+    when 'voting'
+      voting_data(phase, participations)
+    when 'ideation', 'proposals'
+      ideation_data(participations)
+    when 'common_ground'
+      common_ground_data(phase, participations)
+    when 'native_survey'
+      native_survey_data(participations)
+    else
+      {}
+    end
+  end
```

Comment:
Hmm, looking at this, the class is getting very big and I'm also wondering if maybe we're spreading the participation logic too much. Maybe it might be better to have this as a base class and have sub classes that we instantiate that each implement `participation_method_metrics` differently ie:

```
NativeSurveyPhaseInsightsService < IdeationPhaseInsightsService < PhaseInsightsService
```

And then move ALL the logic out of the participation methods - so everything is in one place that is specific to the method. Sorry, that goes against everything I've said before.

Reply:
No need to apologise, that makes perfect sense, and shouldn't take too long to do.

Thanks for the feedback.